From f7ed8c3d7bc689f07ae01c8a1975c9f7dc5def8c Mon Sep 17 00:00:00 2001
From: Jiri Olsa <jolsa@redhat.com>
Date: Mon, 8 Sep 2014 13:06:07 -0400
Subject: [tools] perf/probe: Fix --line option behavior

Message-id: <1410181752-25631-123-git-send-email-jolsa@redhat.com>
Patchwork-id: 91776
O-Subject: [PATCH RHEL7.1 BZ1134356 122/307] perf probe: Fix --line option behavior
Bugzilla: 1134356
RH-Acked-by: Prarit Bhargava <prarit@redhat.com>
RH-Acked-by: Don Zickus <dzickus@redhat.com>

From: Jiri Olsa <jolsa@kernel.org>

Bugzilla: 1134356
https://bugzilla.redhat.com/show_bug.cgi?id=1134356

upstream
========
commit 202c7c123c96a1c193149b7fa2718d7fb143efb2
Author: Namhyung Kim <namhyung@kernel.org>
Date: Tue Apr 1 13:47:57 2014 +0900

description
===========
The commit 5a62257a3ddd1 ("perf probe: Replace line_list with
intlist") replaced line_list to intlist but it has a problem that if a
same line was added again, it'd return -EEXIST rather than 1.

Since line_range_walk_cb() only checks the result being negative, it
resulted in failure or segfault sometimes.
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/tools/perf/util/probe-finder.c b/tools/perf/util/probe-finder.c
index df02386..3bf0c8c 100644
--- a/tools/perf/util/probe-finder.c
+++ b/tools/perf/util/probe-finder.c
@@ -1441,13 +1441,15 @@ static int line_range_walk_cb(const char *fname, int lineno,
          void *data)
 {
  struct line_finder *lf = data;
+ int err;
 
  if ((strtailcmp(fname, lf->fname) != 0) ||
      (lf->lno_s > lineno || lf->lno_e < lineno))
   return 0;
 
- if (line_range_add_line(fname, lineno, lf->lr) < 0)
-  return -EINVAL;
+ err = line_range_add_line(fname, lineno, lf->lr);
+ if (err < 0 && err != -EEXIST)
+  return err;
 
  return 0;
 }
-- 
1.7.1